diesel/src/connection/statement_cache.rs

```diff
@@ -313,15 +310,17 @@ where
     ///
     /// This is an optimization that may skip constructing the query string
     /// twice if it's already part of the current cache key
+    // Note: Intentionally monomorphic over source.
     #[allow(unreachable_pub)]
-    pub fn sql<T: QueryFragment<DB>>(&self, source: &T, backend: &DB) -> QueryResult<Cow<'_, str>> {
+    pub fn sql(&self, source: &dyn QueryFragment<DB>, backend: &DB) -> QueryResult<Cow<'_, str>> {
         match *self {
             StatementCacheKey::Type(_) => Self::construct_sql(source, backend).map(Cow::Owned),
             StatementCacheKey::Sql { ref sql, .. } => Ok(Cow::Borrowed(sql)),
         }
     }
 
-    fn construct_sql<T: QueryFragment<DB>>(source: &T, backend: &DB) -> QueryResult<String> {
+    // Note: Intentionally monomorphic over source.
+    fn construct_sql(source: &dyn QueryFragment<DB>, backend: &DB) -> QueryResult<String> {
```

Comment:
I fear that specifically this one dynamic dispatch here may prevent optimizer from optimizing some specific query fragments into simply functions that return a constant compile-time-computed String...

-> I would be more comfortable if this single part was monomorphized once per statement.

AFAICT that could be done by declaring a new dyn-able trait just here that has this additional construct_sql method inside and is implemented for all QueryFragments (or by adding a construct_sql method to QueryFragment but that seems dirty)

Reply:
I had the same fear, that's why I've requested to run [the benchmarks here](https://github.com/diesel-rs/diesel/pull/3778#issuecomment-1708901348) and they show essentially no regression after that change, even for trivial sqlite queries (selecting a single user from a table that contains a single user). It's important to note that this seems to only happen after the wrapper function around the monomorphize internal `cached_statement_internal` function was removed.

Reply:
> this seems to only happen after

I'm not sure I understand what the *this* refers to in this sencence

Reply:
"This" refers to the benchmark results linked in the sentence before. I've written that because there is another result [here](https://github.com/diesel-rs/diesel/pull/3778#issuecomment-1708721404), which were produced based on an older version that featured the `_internal` method.

Reply:
I had an in detail look at the generated code based on [this example](https://godbolt.org/z/r8MThc59v).  The short version is that we loss the ability to inline the whole ast building into the function that accepts the `&dyn QueryFragment`, but we don't loss the ability to inline `QueryFragments` to each other. In my opinion in does not really matter to have a single virtual call on top as long as the "heavy" work (ast building) is inlined.

This summary is based on the following observations in the linked `godbolt` snippet: The generated assembly for `impl_queryable` (line 308, line 370) and `dyn_queryable` (line 432) is different. Especially the snippet between setting up the fake `AstPass` and the `println!` is much shorter for `dyn_queryable` and contains only a `call` instruction. Also `dyn_queryable` appears only once while `impl_queryable` appears twice (so once for each query). If we now look for `Queryable` impl for `Select<I>` we only see two of these impls in line 174 and line 216. These impls contain two `moveabs` instructions, the first one in the same between both impls (line 185 and line 227), which indicates that this one corresponds to `SELECT *`, while the second one is different (line 193 and line 235). These correspond to the `movabs` instructions in the `Queryable` impls for `From1` and `From2` (line 268 and line 293), which indicates that these impls where inlined. Combined with the fact that there are only two `Queryable` impls for `Select<I>` I conclude that this means that the `dyn_queryable` impl also uses the inline `QueryFragment` impls (as no other one exists).

Based on these results (and the positive benchmark results) I propose to merge this PR as it is.

Reply:
Ah I was essentially scared about the cost of setting up the entire `QueryBuilder` when in fact we only need the output of its `finish`. (In the godbolt example QueryBuilder is only a String).
But it seems currently that is just at worst an integer increment: https://github.com/diesel-rs/diesel/blob/70420834f3831ce08ddf9005d148329105630163/diesel/src/pg/query_builder/mod.rs#L17-L20 and all of the other types used as QueryBuilder hold only a String, and `finish` always just gives back that contained `String`. As long as that is the case I agree that the single dynamic dispatch probably doesn't really matter.
This is what I had in mind though: #3842